File: bit-manipulation/divide-two-integers/divide-two-integers.js
function divide(dividend, divisor) {
  // result in 2,147,483,648 which truncates to 2,147,483,647
  if (dividend === 1 << 31 && divisor === -1) {
    return ~~((1 << 31) - 1);
  }

  let sign = 1;

  if (dividend < 0) {
    sign *= -1;
  }

  if (divisor < 0) {
    sign *= -1;
  }

  let dvd = Math.abs(dividend);
  let dvr = Math.abs(divisor);

  let res = 0;

  while (dvd - dvr >= 0) {
    let x = 0;

    while (~~(dvd - ((dvr << x) << 1)) >= 0) {
      x++;
    }

    res += 1 << x;
    dvd = ~~(dvd - (dvr << x));
  }

  return ~~(res * sign);
}

File: bit-manipulation/divide-two-integers/divide-two-integers.go
package dividetwointegers

import "math"

func divide(dividend int, divisor int) int {
	if dividend == math.MinInt32 && divisor == -1 { // result in 2,147,483,648 which truncates to 2,147,483,647
		return (1 << 31) - 1
	}

	sign := 1

	if dividend < 0 {
		sign *= -1
	}

	if divisor < 0 {
		sign *= -1
	}

	dvd := dividend

	if dvd < 0 {
		dvd *= -1
	}

	dvr := divisor

	if dvr < 0 {
		dvr *= -1
	}

	res := 0

	for dvd-dvr >= 0 {
		x := 0

		for dvd-(dvr<<x<<1) >= 0 {
			x++
		}

		res += 1 << x

		dvd -= dvr << x
	}

	return res * sign
}
